scripts/generate-package-lock.js

```diff
@@ -0,0 +1,190 @@
+          );
+        }
+
+        let meta;
+
+        if (packageNode.isLink) {
+          meta = await resolvePackageMetaForLink(packageNode);
+        } else {
+          meta = Shrinkwrap.metaFromNode(packageNode);
+        }
+
+        packagesMeta.set(metaPath, meta);
+      }
+    }
+  );
+
+  await withProgress(
+    `Writing package lock file to ${path.relative(
+      process.cwd(),
+      path.join(workspacePath, 'package-lock.json')
+    )}`,
+    async () => {
+      // https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json#file-format
+      const packageLock = {
+        name: workspaceName,
+        version: workspaceNode.version,
+        lockfileVersion: 3,
+        packages: Object.fromEntries(packagesMeta)
+      };
+
+      await fs.writeFile(
+        path.join(workspacePath, 'package-lock.json'),
+        JSON.stringify(packageLock, null, 2)
+      );
+    }
+  );
+}
+
+const maybeMissingType = ['optional', 'peer', 'peerOptional'];
+
+function getAllChildrenForNode(nodeOrLink, packages = new Set()) {
+  const node = nodeOrLink.isLink ? nodeOrLink.target : nodeOrLink;
+  for (const edge of node.edgesOut.values()) {
+    const pkg = findPackageNodeRec(edge.name, nodeOrLink);
+    if (!maybeMissingType.includes(edge.type) && !pkg) {
+      throw new Error(
+        `Failed to resolve edge ${edge.name} from package ${
+          node.packageName
+        } at ${node.realpath}:\n\n${JSON.stringify(edge, null, 2)}`
+      );
+    } else if (pkg && !packages.has(pkg)) {
+      packages.add(pkg);
+      getAllChildrenForNode(pkg, packages);
+    }
+  }
+  return packages;
+}
+
+function findPackageNodeRec(packageName, startNode) {
+  const parent = startNode.parent || startNode.top || startNode.root || null;
+  const node = startNode.isLink ? startNode.target : startNode;
+  return startNode.children.has(packageName)
+    ? node.children.get(packageName)
+    : parent && parent !== startNode
+    ? findPackageNodeRec(packageName, parent)
+    : null;
+}
+
+const manifestKeys = [
+  'version',
+  'bin',
+  'license',
+  'engines',
+  'dependencies',
+  'optionalDependencies',
+  '_resolved',
+  '_integrity'
+];
+
+const nodePackageKeys = ['inBundle', 'hasShrinkwrap', 'hasInstallScript'];
+
+/**
+ * Create a shrinkwrap package meta from registry metadata following the
+ * description in npm docs[0] and internal arborist implementation[1] (we can't
+ * use `Shrinkwrap.metaFromNode` directly for LINKs as their metadata will
+ * produce an incorrect shrinkwrap meta)
+ *
+ * [0] - https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json#packages
+ * [1] - https://github.com/npm/arborist/blob/75c785f64bc27f326b645854be0b2607e219f09b/lib/shrinkwrap.js#L107-L146
+ */
+async function resolvePackageMetaForLink(link) {
+  const manifest = await pacote.manifest(`${link.name}@${link.version}`, {
+    // if env is undefined, defaults to https://registry.npmjs.org
+    registry: process.env.npm_config_registry
```

Comment:
Could it be an issue that we don't read `.npmrc`?

Reply:
Good question! So by default when you run an npm script, npm will automatically populate those environment variables (when the values are non-default), that means that all .npmrc configs and other ways to override those will be resolved the same way as npm would resolve them. So running this script through npm scripts will always set this value appropriately and I think to avoid of pulling in all this logic of resolving the value we can skip the use-case of using it not from npm scripts. What do you think?

Reply:
I'll merge this as-is, but will follow-up if we think we need to make this logic here more sturdy 👍